src/modlunky2/ui/trackers/runstate.py

```diff
@@ -630,40 +661,54 @@ def update(self, game_state: State):
         hud_flags = game_state.hud_flags
         presence_flags = game_state.presence_flags
         self.update_global_state(game_state)
-        self.update_on_level_start()
+        self.update_on_level_start(self.world, self.theme, self.ropes)
         self.update_player_item_types(game_state.instance_id_to_pointer, player)
-        self.update_final_death()
+        self.update_final_death(state, self.player_item_types)
 
-        self.update_score_items()
+        self.update_score_items(self.world, self.player_item_types)
 
         # Check Modifiers
         self.update_pacifist(run_recap_flags)
         self.update_no_gold(run_recap_flags)
-        self.update_no_tp()
-        self.update_eggplant()
+        self.update_no_tp(self.player_item_types)
+        self.update_eggplant(self.world, self.player_item_types)
 
         # Check Category Criteria
         overlay = player.overlay
 
         # Low%
-        self.update_has_mounted_tame(overlay)
-        self.update_starting_resources(player, inventory)
-        self.update_status_effects()
+        self.update_has_mounted_tame(self.theme, overlay)
+        self.update_starting_resources(player, self.player_state, inventory)
+        self.update_status_effects(self.player_state, self.player_item_types)
         self.update_had_clover(hud_flags)
-        self.update_wore_backpack()
-        self.update_held_shield()
-        self.update_has_non_chain_powerup()
-        self.update_attacked_with(layer, presence_flags)
-        self.update_attacked_with_throwables()
+        self.update_wore_backpack(self.player_item_types)
+        self.update_held_shield(self.player_item_types)
+        self.update_has_non_chain_powerup(self.player_item_types)
+        self.update_attacked_with(
+            last_state,
+            state,
+            layer,
+            self.world,
+            self.level,
+            self.theme,
+            presence_flags,
+            self.player_item_types,
+        )
+        self.update_attacked_with_throwables(
+            self.player_last_state,
+            self.player_state,
+            self.player_last_item_types,
+            self.player_item_types,
+        )
 
         # Chain
-        self.update_chain()
-        self.update_has_chain_powerup()
-        self.update_is_chain()
+        self.update_chain(self.player_item_types)
+        self.update_has_chain_powerup(self.chain_status, self.player_item_types)
+        self.update_is_chain(self.world, self.level, self.theme, self.win_state)
 
-        self.update_millionaire(game_state, inventory)
+        self.update_millionaire(game_state, inventory, self.player_item_types)
 
-        self.update_terminus()
+        self.update_terminus(self.world, self.theme, self.win_state)
```

Comment:
Definitely feels weird to be passing instance attributes to methods as parameters. Need to think about this one a bit more. It's kind of hard to reason about because it feels like a static method.

Reply:
Yeah, my intention is to stop storing `State` stuff in `RunState`. I want to shift to either passing `State` (or other dataclass) itself, or pass the values separately but just fetch them from `State`. I'm leaning towards the former for convenience.